python/cudf_polars/cudf_polars/experimental/join.py

```diff
@@ -0,0 +1,275 @@
+
+    from cudf_polars.dsl.expr import NamedExpr
+    from cudf_polars.dsl.ir import IR
+    from cudf_polars.experimental.parallel import LowerIRTransformer
+
+
+def _make_hash_join(
+    ir: Join,
+    output_count: int,
+    partition_info: MutableMapping[IR, PartitionInfo],
+    left: IR,
+    right: IR,
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    shuffle_options: dict[str, Any] = {}
+
+    if not (
+        partition_info[left].partitioned_on == ir.left_on
+        and partition_info[left].count == output_count
+    ):
+        left = Shuffle(
+            left.schema,
+            ir.left_on,
+            shuffle_options,
+            left,
+        )
+        partition_info[left] = PartitionInfo(
+            count=output_count,
+            partitioned_on=ir.left_on,
+        )
+
+    if not (
+        partition_info[right].partitioned_on == ir.right_on
+        and partition_info[right].count == output_count
+    ):
+        right = Shuffle(
+            right.schema,
+            ir.right_on,
+            shuffle_options,
+            right,
+        )
+        partition_info[right] = PartitionInfo(
+            count=output_count,
+            partitioned_on=ir.right_on,
+        )
+
+    new_node = ir.reconstruct([left, right])
+
+    partitioned_on: tuple[NamedExpr, ...] = ()
+    if ir.left_on == ir.right_on or (ir.options[0] in ("left", "semi", "anti")):
+        partitioned_on = ir.left_on
+    elif ir.options[0] == "right":  # pragma: no cover
+        partitioned_on = ir.right_on
+    partition_info[new_node] = PartitionInfo(
+        count=output_count,
+        partitioned_on=partitioned_on,
+    )
+
+    return new_node, partition_info
+
+
+def _make_bcast_join(
+    ir: Join,
+    output_count: int,
+    partition_info: MutableMapping[IR, PartitionInfo],
+    left: IR,
+    right: IR,
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    how = ir.options[0]
+    if how != "inner":
+        shuffle_options: dict[str, Any] = {}
+        left_count = partition_info[left].count
+        right_count = partition_info[right].count
+        if left_count >= right_count:
+            right = Shuffle(
+                right.schema,
+                ir.right_on,
+                shuffle_options,
+                right,
+            )
+            partition_info[right] = PartitionInfo(
+                count=right_count,
+                partitioned_on=ir.right_on,
+            )
+        else:
+            left = Shuffle(
+                left.schema,
+                ir.left_on,
+                shuffle_options,
+                left,
+            )
+            partition_info[left] = PartitionInfo(
+                count=left_count,
+                partitioned_on=ir.left_on,
+            )
+
+    new_node = ir.reconstruct([left, right])
+    partition_info[new_node] = PartitionInfo(count=output_count)
+    return new_node, partition_info
+
+
+@lower_ir_node.register(Join)
+def _(
```

Comment:
So whether or not we choose to do a broadcast join or hash join depends on
- Whether the "large table" is shuffled 
- bcast_count
- join type

Is is worth consolidating this into `_should_broadcast_join` to its clearer why we pick either?

Reply:
Yeah, I think `_should_broadcast_join` is a good idea. Just added it.